File: sdk/python/tellama.py
"""Dependency-free client for the Tellama local AI server.

SPDX-License-Identifier: Apache-2.0
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Iterator, Sequence
from urllib.error import HTTPError, URLError
from urllib.parse import urlparse
from urllib.request import Request, urlopen


@dataclass
class TellamaError(Exception):
    """An HTTP or Tellama API error with a safe, non-secret description."""

    message: str
    status: int | None = None
    code: str | None = None

    def __str__(self) -> str:
        details = [part for part in (self.code, str(self.status) if self.status else None) if part]
        prefix = " / ".join(details)
        return f"{prefix}: {self.message}" if prefix else self.message


class TellamaClient:
    """Small synchronous client for Ollama and OpenAI-compatible streaming routes."""

    def __init__(self, base_url: str, api_key: str | None = None, timeout: float = 180.0):
        parsed = urlparse(base_url)
        if parsed.scheme not in {"http", "https"} or not parsed.hostname:
            raise ValueError("base_url must be an absolute http(s) URL")
        if timeout <= 0:
            raise ValueError("timeout must be greater than zero")
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key.strip() if api_key else None
        self.timeout = timeout

    def list_models(self) -> list[dict[str, Any]]:
        """Return Ollama-compatible model descriptors from GET /api/tags."""
        with self._open("GET", "/api/tags") as response:
            payload = json.load(response)
        models = payload.get("models", [])
        if not isinstance(models, list):
            raise TellamaError("The server returned an invalid model list")
        return models

    def ollama_chat(
        self,
        model: str,
        messages: Sequence[dict[str, str]],
    ) -> Iterator[str]:
        """Yield visible text chunks from POST /api/chat (NDJSON)."""
        body = {"model": model, "messages": list(messages), "stream": True}
        with self._open("POST", "/api/chat", body) as response:
            for raw_line in response:
                line = raw_line.decode("utf-8").strip()
                if not line:
                    continue
                payload = json.loads(line)
                chunk = payload.get("message", {}).get("content", "")
                if chunk:
                    yield chunk
                if payload.get("done") is True:
                    return

    def openai_chat(
        self,
        model: str,
        messages: Sequence[dict[str, str]],
    ) -> Iterator[str]:
        """Yield text chunks from POST /v1/chat/completions (SSE)."""
        body = {"model": model, "messages": list(messages), "stream": True}
        with self._open("POST", "/v1/chat/completions", body) as response:
            for raw_line in response:
                line = raw_line.decode("utf-8").strip()
                if not line.startswith("data:"):
                    continue
                data = line.removeprefix("data:").strip()
                if data == "[DONE]":
                    return
                payload = json.loads(data)
                choices = payload.get("choices", [])
                if choices:
                    chunk = choices[0].get("delta", {}).get("content", "")
                    if chunk:
                        yield chunk

    def _open(self, method: str, path: str, body: dict[str, Any] | None = None):
        headers = {"Accept": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        data = None
        if body is not None:
            data = json.dumps(body, separators=(",", ":")).encode("utf-8")
            headers["Content-Type"] = "application/json"
        request = Request(f"{self.base_url}{path}", data=data, headers=headers, method=method)
        try:
            return urlopen(request, timeout=self.timeout)
        except HTTPError as error:
            raise self._http_error(error) from None
        except (URLError, TimeoutError, OSError) as error:
            reason = getattr(error, "reason", error)
            raise TellamaError(f"Could not reach Tellama: {reason}") from None

    @staticmethod
    def _http_error(error: HTTPError) -> TellamaError:
        try:
            payload = json.loads(error.read().decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError):
            payload = {}
        body = payload.get("error") if isinstance(payload.get("error"), dict) else payload
        return TellamaError(
            message=body.get("message") or "Tellama request failed",
            status=error.code,
            code=body.get("code"),
        )


__all__ = ["TellamaClient", "TellamaError"]

File: sdk/javascript/tellama.mjs
/** Dependency-free Tellama client. SPDX-License-Identifier: Apache-2.0 */
export class TellamaError extends Error {
  constructor(message, { status = null, code = null } = {}) {
    super(message);
    this.name = "TellamaError";
    this.status = status;
    this.code = code;
  }
}

export class TellamaClient {
  constructor({ baseUrl, apiKey = null, timeoutMs = 180_000 }) {
    const parsed = new URL(baseUrl);
    if (!["http:", "https:"].includes(parsed.protocol)) {
      throw new TypeError("baseUrl must use http or https");
    }
    if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
      throw new TypeError("timeoutMs must be a positive finite number");
    }
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.apiKey = apiKey?.trim() || null;
    this.timeoutMs = timeoutMs;
  }

  async listModels() {
    const response = await this.#request("/api/tags");
    const payload = await response.json();
    if (!Array.isArray(payload.models)) {
      throw new TellamaError("The server returned an invalid model list");
    }
    return payload.models;
  }

  async *ollamaChat({ model, messages }) {
    const response = await this.#request("/api/chat", {
      method: "POST",
      body: JSON.stringify({ model, messages, stream: true }),
    });
    for await (const line of responseLines(response, this.timeoutMs)) {
      if (!line) continue;
      const payload = JSON.parse(line);
      const content = payload.message?.content || payload.response || "";
      if (content) yield content;
      if (payload.done === true) return;
    }
  }

  async *openAIChat({ model, messages }) {
    const response = await this.#request("/v1/chat/completions", {
      method: "POST",
      body: JSON.stringify({ model, messages, stream: true }),
    });
    for await (const line of responseLines(response, this.timeoutMs)) {
      if (!line.startsWith("data:")) continue;
      const data = line.slice(5).trim();
      if (data === "[DONE]") return;
      const payload = JSON.parse(data);
      const content = payload.choices?.[0]?.delta?.content || "";
      if (content) yield content;
    }
  }

  async #request(path, options = {}) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    const headers = { Accept: "application/json", ...(options.headers || {}) };
    if (options.body) headers["Content-Type"] = "application/json";
    if (this.apiKey) headers.Authorization = `Bearer ${this.apiKey}`;
    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        ...options,
        headers,
        signal: controller.signal,
      });
      if (!response.ok) {
        let payload = {};
        try {
          payload = await response.json();
        } catch {
          // Keep the public error free of response internals.
        }
        const body = payload.error && typeof payload.error === "object" ? payload.error : payload;
        throw new TellamaError(
          body.message || "Tellama request failed",
          { status: response.status, code: body.code || null },
        );
      }
      return response;
    } catch (error) {
      if (error instanceof TellamaError) throw error;
      const message = error?.name === "AbortError"
        ? "Tellama request timed out"
        : "Could not reach Tellama";
      throw new TellamaError(message);
    } finally {
      clearTimeout(timer);
    }
  }
}

async function* responseLines(response, timeoutMs) {
  if (!response.body) throw new TellamaError("Tellama returned an empty response");
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    while (true) {
      const { value, done } = await readWithTimeout(reader, timeoutMs);
      buffer += decoder.decode(value || new Uint8Array(), { stream: !done });
      const lines = buffer.split(/\r?\n/);
      buffer = lines.pop() || "";
      for (const line of lines) yield line;
      if (done) break;
    }
    if (buffer) yield buffer;
  } finally {
    reader.releaseLock();
  }
}

async function readWithTimeout(reader, timeoutMs) {
  let timer;
  try {
    return await Promise.race([
      reader.read(),
      new Promise((_, reject) => {
        timer = setTimeout(
          () => reject(new TellamaError("Tellama stream timed out")),
          timeoutMs,
        );
      }),
    ]);
  } finally {
    clearTimeout(timer);
  }
}
